src/metadataExt.ts

```diff
@@ -119,55 +193,37 @@ export class MetadataModuleExt {
         return `Copy of ${originalName}`;
     }
 
-    private async updateWidgets(dashboardUri: string, uriTranslator: UriTranslator): Promise<any> {
-        const { analyticalDashboard } = await this.metadataModule.getObjectDetails(dashboardUri);
-
-        return analyticalDashboard.content.widgets.map((uri: string) => {
-            return uriTranslator(uri);
-        });
-    }
-
-    private async updateLayout(dashboardUri: string, uriTranslator: UriTranslator): Promise<any> {
-        const { analyticalDashboard } = await this.metadataModule.getObjectDetails(dashboardUri);
-        return cloneDeepWith({ ...analyticalDashboard.content.layout.fluidLayout }, key => {
-            const uri = key.uri;
-            if (uri === undefined) {
-                return;
-            }
-            return {
-                ...key,
-                uri: uriTranslator(uri),
-            };
-        });
-    }
-
     private async duplicateOrKeepKpis(
         projectId: string,
-        objsFromDashboard: any,
+        objsFromDashboard: any[],
         options: ICopyDashboardOptions,
-    ): Promise<any> {
-        const uriMap: any = new Map();
-        objsFromDashboard
-            .filter((obj: any) => this.unwrapObj(obj).meta.category === "kpi")
-            .map(async (kpiWidget: any) => {
-                const { kpi } = kpiWidget;
-                if (options.copyKpi) {
-                    const newUriKpiObj = (await this.metadataModule.createObject(projectId, kpiWidget)).kpi
-                        .meta.uri;
-                    uriMap.set(kpi.meta.uri, newUriKpiObj);
-                } else {
-                    uriMap.set(kpi.meta.uri, kpi.meta.uri);
-                }
-            });
+    ): Promise<Map<string, string>> {
+        const uriMap: Map<string, string> = new Map();
+        await Promise.all(
+            objsFromDashboard
+                .filter((obj: any) => this.unwrapObj(obj).meta.category === "kpi")
+                .map(async (kpiWidget: any) => {
+                    const { kpi }: { kpi: IKPI } = kpiWidget;
+                    if (options.copyKpi || typeof options.copyKpi === "undefined") {
+                        const newUriKpiObj: string = (await this.metadataModule.createObject(
+                            projectId,
+                            kpiWidget,
+                        )).kpi.meta.uri;
+                        uriMap.set(kpi.meta.uri as string, newUriKpiObj);
+                    } else {
+                        uriMap.set(kpi.meta.uri as string, kpi.meta.uri as string);
```

Comment:
Detail: You probably do not need the `else` as the translator returns the original uri anyway if it is not in the map. This means that this whole `await Promise.all` is not needed in case `options.copyKpi === false`, right?

Reply:
I think we still need `await Promise.all` in case there are multiple KPIs in the dashboard.

Reply:
Yes, we do but we don't need to even start it, if `options.copyKpi === false`, right?

Reply:
Got your point. Thanks for the explanation.